lib/pbench/agent/tool_meister_client.py

```diff
@@ -0,0 +1,227 @@
+# -*- mode: python -*-
+
+"""tool_meister_client API
+
+The Tool Meister Client API encapsulates the steps required to send a message
+to the Tool Data Sink with the action to be taken by the Tool Meisters,
+processing the response that arrives from the Tool Data Sink for success or
+failure.
+
+A client is really only sending a message to the Tool Data Sink.  The Tool
+Data Sink, in turn, coordinates all the behaviors related to the given action
+for all the Tool Meisters, including forwarding the action payload to the Tool
+Meisters.
+"""
+
+import json
+import logging
+
+import redis
+
+from pbench.agent.constants import (
+    tm_channel_suffix_to_client,
+    tm_channel_suffix_from_client,
+    api_tm_allowed_actions,
+)
+from pbench.agent.redis import RedisChannelSubscriber
+
+
+class Client:
+    """Context manager Tool Meister client.
+
+    The constructor records the necessary client information for the context
+    manager "enter" and "exit" methods to operate.
+    """
+
+    def __init__(
+        self,
+        redis_server=None,
+        redis_host=None,
+        redis_port=None,
+        channel_prefix=None,
+        to_client_chan=None,
+        logger=None,
+    ):
+        """Construct a Tool Meister "client" object, given the host and port
+        of a Redis server.  The caller can optionally provide a logger to be
+        used.
+
+        This constructor does not contact the Redis server, it just records
+        the information for where that Redis server is.
+
+        :redis_server:   - (optional) a previously construct Redis server
+                           client object
+        :redis_host:     - (optional) the IP or host name of the Redis server
+                           to use
+        :redis_port:     - (optional) the port on which the Redis server on
+                           the given host name is listening
+        :channel_prefix: - (required) the prefix string to use for the channel
+                           name
+        :to_client_chan: - (optional) a previously constructed
+                           RedisChannelSubscriber object (the caller ensures
+                           it is for the same Redis server client object
+                           given)
```

Comment:
Document that we will now own this channel object and will close it when we are done.

Reply:
Tracked in https://github.com/distributed-system-analysis/pbench/issues/2158.